Improve crafting recipe documentation (#12806)
minetest/minetest: doc/lua_api.txt

Crafting recipes
----------------

All crafting recipes take one or more items as input and have an
optional output.

> Crafting needs inventory slots. From the crafting inputs, only one

Reviewer: > Crafting needs inventory slots. 

What do you mean by that? `minetest.get_craft_result` for example does not need inventory slots, and the commonly known act of "crafting" obviously needs space to work with.

> All crafting recipes take one or more items as input and have an
> optional output.
> Crafting needs inventory slots. From the crafting inputs, only one
> item per slot can be used per crafting, while the output is a single
> item stack of arbitrary size.

Reviewer: ```suggestion
Crafting converts one or more inputs to one output ItemStack of arbitrary count
(including empty). The conversion reduces each input ItemStack by one (1).
```

> Crafting recipe definitions are always a table with a number of
> parameters (listed below). They must be used as input parameter
> for `minetest.register_craft`.

Reviewer: ```suggestion
Craft recipes are registered by `minetest.register_craft`. The accepted
formats are listed below..
```

> Input items can be specified via itemname, that is, without itemcount
> or wear. (The item count is always 1.)
> 
> Input slots may alternatively accept an item from a group, with a syntax like
> `group:<group_name>`. See "Groups in crafting recipes" for details.

Reviewer: ```suggestion
Recipe input items can either be specified by item name (item count = 1)
or by group (see "Groups in crafting recipes" for details).
```

The following sections describe the types and syntaxes of recipes.

### Shaped

This is the default recipe type (when no `type` is specified).

A shaped recipe takes one or multiple items as input and has
a single item stack as output. The input items must be specified
in a 2-dimensional matrix (see parameters below) to specify the
exact arrangement (the "shape") in which the player must place them
in the crafting grid.

For example, for a 3x3 recipe, the `recipes` table must have
3 rows and 3 columns.

> For the player to be able to craft a shaped recipe, they must
> have access to a crafting grid that has the same or larger
> width and height as the recipe. E.g. a crafting grid of
> size 2x2 can only craft recipes of size 2x2, 2x1, 1x2 and 1x1,
> but not recipes of size 3x3, 3x2, 2x3, etc.

Reviewer: This largely seems superfluous/obvious.
```suggestion
In order to craft the recipe, the players' crafting grid must have matching or larger dimensions.
```

Parameters:

* `type = "shaped"`: (optional) specifies recipe type as shaped

> * `output`: Itemstring of output itemstack (item counts >= 1 are allowed)

Reviewer: ```suggestion
* `output`: Itemstring of the output ItemStack
```
Crafting behaviour (count >= 1) is already documented above.

* `recipe`: A 2-dimensional matrix of items, with a width *w* and height *h*.
    * *w* and *h* are chosen by you, they don't have to be equal but must be at least 1
    * The matrix is specified as a table containing tables containing itemnames
    * The inner tables are the rows. There must be *h* tables, specified from the top to the bottom row
    * Values inside of the inner table are the columns.

>       Each inner table must contain a list of *w* items, specified from left to right

Reviewer: I suppose a commented code example would be more helpful than this text.

    * Empty slots *must* be filled with the empty string

> * `replacements`: Optionally allows you to replace input items with some other items
>       when something is crafted

Reviewer: ```suggestion
* `replacements`: (optional) replacement rules of input items on craft
```

    * Provided as a list of item pairs of the form `{ old_item, new_item }` where
      `old_item` is the input item to replace (same syntax as for a regular input
      slot; groups are allowed) and `new_item` is an itemstring for the item stack
      it will become

>     * When the output is crafted, Minetest iterates through the list

Reviewer: What if the input stack has a count larger than 1? What's replaced?

      of input items if the crafting grid. For each input item stack, it checks if
      it matches with an `old_item` in the item pair list.
        * If it matches, the item will be replaced. Also, this item pair
          will *not* be applied again for the remaining items
        * If it does not match, the item is consumed (reduced by 1) normally
    * Item replacements will appear either in the crafting grid, the
      player inventory or will be dropped as item, depending on inventory space

#### Examples

A typical shaped recipe:

    -- Stone pickaxe
    {
        output = "example:stone_pickaxe",

>         -- This is a 3x3 recipe which needs 3 cobblestone in the 1st row,
>         -- and 1 stick in the horizontal middle in each of the 2nd and 3nd row.
>         -- The 4 remaining slots have to be empty.
>         recipe = {
>             {"example:cobblestone", "example:cobblestone", "example:cobblestone"},
>             {"", "example:stick", ""},
>             {"", "example:stick", ""},

Reviewer: ```suggestion
        -- A 3x3 recipe containing 4 empty slots.
       recipe = {
            {"mymod:stone", "mymod:stone", "mymod:stone"}, -- row 1
            {"",            "mymod:stick", ""           }, -- row 2
            {"",            "mymod:stick", ""           }, -- row 3
            -- ^ column 1   ^ column 2     ^ column 3
          
```

            -- Without these repetitions, only the first
            -- water bucket would be replaced.
            {"example:water_bucket", "example:empty_bucket"},
            {"example:water_bucket", "example:empty_bucket"},
            {"example:water_bucket", "example:empty_bucket"},
            {"example:water_bucket", "example:empty_bucket"},
        },
    }

Complex replacement example 2:

    -- Magic book:
    -- 3 magic orbs + 1 book crafts a magic book,
    -- and the orbs will be replaced with 3 different runes.
    {
        output = "example:magic_book",
        -- 3x2 recipe
        recipe = {
            -- 3 items in the group `magic_orb` on top of a book in the middle
            {"group:magic_orb", "group:magic_orb", "group:magic_orb"},
            {"", "example:book", ""},
        },
        -- When the book is crafted, the 3 magic orbs will be turned into
        -- 3 runes: ice rune, earth rune and fire rune (from left to right)
        replacements = {
            {"group:magic_orb", "example:ice_rune"},
            {"group:magic_orb", "example:earth_rune"},
            {"group:magic_orb", "example:fire_rune"},
        },
    }

### Shapeless

> Same as shaped recipe, except the arrangement of items in the input
> crafting grid does not matter. Also, the recipe is a list of items instead
> of a 2D matrix.

Reviewer: ```suggestion
Takes an 1-dimensional array of at least one input item. The order
of input items does not matter.
```

> For the player to be able to craft a shapeless recipe, they must have
> access to a crafting grid that has at least as many slots
> as the number of required items. E.g. a crafting grid of size 3x3 can
> craft shapeless recipes with up to 9 input items. The width of height of
> the grid do *not* matter.

Reviewer: This largely seems superfluous/obvious.
```suggestion
In order to craft the recipe, the players' crafting grid must have matching or
larger *count* of slots. The grid dimensions do not matter.
```
